File: src/main/java/com/softwareloop/glo/RomProcessor.java
package com.softwareloop.glo;

import com.softwareloop.glo.model.RomSummary;
import lombok.Getter;
import lombok.Setter;
import lombok.SneakyThrows;

import javax.xml.bind.DatatypeConverter;
import java.io.BufferedInputStream;
import java.io.InputStream;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class RomProcessor {

    //--------------------------------------------------------------------------
    // Constants
    //--------------------------------------------------------------------------

    //--------------------------------------------------------------------------
    // Fields
    //--------------------------------------------------------------------------

    private final DatStore datStore;

    @Getter
    @Setter
    boolean renameEnabled;

    @Getter
    private int nProcessedFiles;
    @Getter
    private int nMatchedFiles;
    @Getter
    private int nUnmatchedFiles;
    @Getter
    private int nRenamedFiles;

    //--------------------------------------------------------------------------
    // Constructors
    //--------------------------------------------------------------------------

    public RomProcessor(DatStore datStore) {
        this.datStore = datStore;
    }

    //--------------------------------------------------------------------------
    // Public methods
    //--------------------------------------------------------------------------

    @SneakyThrows
    public void processDir(Path romDir) {
        List<String> unmatched = new ArrayList<>();
        List<String> fileNames = new ArrayList<>();
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(romDir)) {
            for (Path romFile : stream) {
                String fileName = romFile.getFileName().toString();
                if (Files.isRegularFile(romFile) && !fileName.startsWith(".")) {
                    fileNames.add(fileName);
                }
            }
        }
        fileNames.sort(String::compareToIgnoreCase);
        for (String fileName : fileNames) {
            boolean matched = processRom(romDir, fileName);
            if (!matched) {
                unmatched.add(fileName);
            }
        }
        if (!unmatched.isEmpty()) {
            Log.info("Unmatched files:");
            for (String romFile : unmatched) {
                Log.info(romFile);
            }
        }
    }

    @SneakyThrows
    private boolean processRom(
            Path romDir,
            String fileName
    ) {
        nProcessedFiles++;
        Path romFile = romDir.resolve(fileName);
        String md5 = computeMd5(romFile);
        List<RomSummary> romSummaries = datStore.getRomSummaryByMd5(md5);
        if (romSummaries == null) {
            Log.debug("No match found");
            nUnmatchedFiles++;
            return false;
        }
        nMatchedFiles++;
        if (renameEnabled) {
            Set<String> newFileNames = new HashSet<>();
            for (RomSummary romSummary : romSummaries) {
                String newFileName = romSummary.getRomName();
                newFileNames.add(newFileName);
            }
            if (newFileNames.size() == 1) {
                String newFileName = newFileNames.iterator().next();
                if (fileName.equals(newFileName)) {
                    Log.debug("Name matches dat entry: {}", fileName);
                } else {
                    Log.info("Renaming {} -> {}", fileName, newFileName);
                    Path newRomFile = romDir.resolve(newFileName);
                    Files.move(romFile, newRomFile, StandardCopyOption.REPLACE_EXISTING);
                    nRenamedFiles++;
                }
            } else {
                if (newFileNames.contains(fileName)) {
                    Log.debug("Name matches dat entry: {}", fileName);
                } else {
                    Log.info("Skipping {} - multiple matching rom names:", fileName);
                    for (RomSummary romSummary : romSummaries) {
                        String newFileName = romSummary.getRomName();
                        Log.info("    {} [{}]", newFileName, romSummary.getDatName());
                    }
                }
            }
        } else {
            Log.info(fileName);
            for (RomSummary romSummary : romSummaries) {
                String newFileName = romSummary.getRomName();
                Log.info("    {} [{}]", newFileName, romSummary.getDatName());
            }
        }
        return true;
    }

    //--------------------------------------------------------------------------
    // Private methods
    //--------------------------------------------------------------------------

    @SneakyThrows
    private String computeMd5(Path romFile) {
        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] buffer = new byte[1024];
        try (InputStream is = new BufferedInputStream(Files.newInputStream(romFile))) {
            int len;
            while ((len = is.read(buffer)) > 0) {
                md.update(buffer, 0, len);
            }
            byte[] digest = md.digest();
            return DatatypeConverter.printHexBinary(digest).toUpperCase();
        }
    }

    public void printStats() {
        Log.info("\nFile stats:");
        Log.info("Processed: {}", nProcessedFiles);
        Log.info("Matched  : {}", nMatchedFiles);
        Log.info("Unmatched: {}", nUnmatchedFiles);
        Log.info("Renamed  : {}", nRenamedFiles);
    }
}

File: src/main/java/com/softwareloop/glo/RomProcessor.kt
package com.softwareloop.glo

import com.softwareloop.glo.dat.DatStore
import com.softwareloop.glo.dat.RomSummary
import com.softwareloop.glo.util.getBaseFilename
import com.softwareloop.glo.util.getFilenameExtension
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption


class RomProcessor(
    val datStore: DatStore,
    val indentation: String = ""
) {

    //--------------------------------------------------------------------------
    // Properties
    //--------------------------------------------------------------------------

    var renameEnabled = false
    var unzipEnabled = false

    private val matchedFiles: MutableMap<String, RomSummary> = HashMap()
    private val unmatchedFiles: MutableList<String> = ArrayList()
    private var nRenamedFiles = 0
    private var nUnzippedFiles = 0

    //--------------------------------------------------------------------------
    // Public methods
    //--------------------------------------------------------------------------

    fun processDir(romDir: Path) {
        val fileNames: MutableList<String> = ArrayList()
        Files.newDirectoryStream(romDir).use { stream ->
            for (romFile in stream) {
                val fileName = romFile.fileName.toString()
                if (Files.isRegularFile(romFile) && !fileName.startsWith(".")) {
                    fileNames.add(fileName)
                }
            }
        }
        fileNames.sortWith { obj: String, str: String? -> obj.compareTo(str!!, ignoreCase = true) }
        for (fileName in fileNames) {
            Log.info("%s%s", indentation, fileName)
            val fileExtension = fileName.getFilenameExtension().lowercase()
            if ("zip".equals(fileExtension)) {
                processZip(romDir, fileName)
            } else {
                processRom(romDir, fileName)
            }
        }
    }

    private fun processZip(
        romDir: Path,
        zipName: String
    ) {
        val zipFile = romDir.resolve(zipName)
        val zipRomSummary = RomSummary()
        FileSystems.newFileSystem(zipFile, ClassLoader.getSystemClassLoader()).use {
            for (rootDirectory in it.rootDirectories) {
                val zipRomProcessor = RomProcessor(datStore, indentation + "    ")
                zipRomProcessor.renameEnabled = renameEnabled
                zipRomProcessor.processDir(rootDirectory)
                for ((romName, romSummary) in zipRomProcessor.matchedFiles) {
                    zipRomSummary.addAll(romSummary)
                    if (unzipEnabled) {
                        val fromPath = rootDirectory.resolve(romName)
                        val toPath = romDir.resolve(romName)
                        Log.info("%s    Extracting to: %s", indentation, romName)
                        Files.copy(fromPath, toPath, StandardCopyOption.REPLACE_EXISTING)
                        nUnzippedFiles++
                    }
                }
            }
        }
        if (zipRomSummary.isEmpty()) {
            Log.debug("%s    No match found", indentation)
            unmatchedFiles.add(zipName)
            return
        }
        val finalFileName = if (renameEnabled) {
            rename(zipRomSummary, romDir, zipName, true) { commonName ->
                val baseName = commonName.getBaseFilename()
                "$baseName.zip"
            }
        } else zipName
        matchedFiles[finalFileName] = zipRomSummary
    }

    private fun processRom(
        romDir: Path,
        fileName: String
    ) {
        val romFile = romDir.resolve(fileName)
        val md5s = Md5Processor.computeMd5s(romFile)
        var romSummary: RomSummary? = null
        for (md5 in md5s) {
            romSummary = datStore.getRomSummaryByMd5(md5)
            if (romSummary != null) {
                break
            }
        }
        if (romSummary == null) {
            Log.debug("%s    No match found", indentation)
            unmatchedFiles.add(fileName)
            return
        }
        for (romEntry in romSummary.romEntries) {
            val newFileName = romEntry.romName
            Log.info("%s    %s [%s]", indentation, newFileName, romEntry.datName)
        }
        val finalFileName = if (renameEnabled) {
            rename(romSummary, romDir, fileName, false) { s -> s }
        } else fileName
        matchedFiles[finalFileName] = romSummary
    }

    private fun rename(
        romSummary: RomSummary,
        romDir: Path,
        fileName: String,
        ignoreExtension: Boolean,
        newFileNameSupplier: (String) -> String
    ): String {
        return if (romSummary.containsRomName(fileName, ignoreExtension)) {
            Log.debug("%s    No need to rename", indentation)
            fileName
        } else {
            val commonName = romSummary.getCommonName()
            if (commonName == null) {
                Log.info("%s    Multiple matching rom names. Not renaming.", indentation)
                fileName
            } else {
                val newFileName = newFileNameSupplier.invoke(commonName)
                rename(romDir, fileName, newFileName)
                newFileName
            }
        }
    }

    private fun rename(
        romDir: Path,
        fileName: String,
        newFileName: String
    ) {
        Log.info("%s    Renaming to: %s", indentation, newFileName)
        val romFile = romDir.resolve(fileName)
        val newRomFile = romDir.resolve(newFileName)
        Files.move(romFile, newRomFile, StandardCopyOption.REPLACE_EXISTING)
        nRenamedFiles++
    }

    fun printUnmatched() {
        Log.info("\n%sUnmatched files:", indentation)
        if (unmatchedFiles.isEmpty()) {
            Log.info("%sNo unmatched files", indentation)
        } else {
            for (romFile in unmatchedFiles) {
                Log.info("%s%s", indentation, romFile)
            }
        }
    }

    fun printStats() {
        with(Log) {
            info("\nFile stats:")
            info("Processed: %s", matchedFiles.size + unmatchedFiles.size)
            info("Matched  : %s", matchedFiles.size)
            info("Unmatched: %s", unmatchedFiles.size)
            info("Renamed  : %s", nRenamedFiles)
        }
    }

}